### DISCUSSION TITLE
zh/Q-and-A/answer/question3

### DISCUSSION THREAD

**Discussion by giscus** [vuepress-theme-hope/giscus-discussions#916] [Announcements]
# zh/Q-and-A/answer/question3

为什么我的输出总不对 问题描述： 这道题不就是分离整数，再判断是不是2吗？硬控我15分钟了

https://cygnomatic-controller.github.io/zh/Q-and-A/answer/question3.html

**Comment by ars0819** [comment]
这个输入可能是多位数，你这个 if(i>=10 && i/10 == 2)
          {
               num++;
          }只检查了十位数字是不是2没有检查更高位。可以尝试用循环

**Comment by MareDevi** [comment]
分离整数是逐位分离的
Op的算法有问题吧

**Comment by uniformshang** [comment]
你可以在循环里再加一个while循环，比如while（i>0){if(i%10==2){num++};i=i/10},这样应该就避免了多位整数的问题

**Comment by lyricsz1115** [comment]
我觉得应该先把区间内的每一个数字用取余和除分离成两个数字然后分别检验带不带2的 这个i/10==2只有20没有21到29之间的数 循环的条件不对

**Reply by lyricsz1115** [reply]
可以用while把21到29的数num++一下

**Comment by WangJie-Mant** [comment]
你的思路是如果是个位单个的2的话就从%2进num++，如果十位的2的话就从/10进num++，但当i为222这样的数值的时候，i/10==2这样的逻辑判断就不再可靠，导致你的num累加的结果不正确。
更好的方法是试图用自除和取余分别取出每一位的同时把i缩小，所以可以试试两层循环。第一层用来i++推进程序，第二层来记录i每位上的数字，再判断每位数字是否为2

**Comment by wshy531** [comment]
注意读题（捂脸），这个真的很重要，经常因为一个大于和大于等于的区别然后WA
习题的最下面 1≤L≤R≤100000
好像确实用while感觉好一点，就是可能反复赋值，逻辑得整清楚（可以多调试监视几遍
for循环的话，可能就得提前确定好几位数，然后看循环几次，要不就结束条件不写，最后break也行，但感觉有点多余
（可能有说的不对的，仅供参考（再次捂脸

**Comment by keee9490** [comment]
你的程序并不能保证多位为2的数据，例如22中的每个2都被统计到。其次如果出现3位或更高位，程序无法判断出2的个数。
所以建议你在for循环内部嵌套while循环，通过先%10判断当前个位是否有2，再/10将个位清除（相当于之前的10位变成个位），直到将每位上的2都统计到。
你也可以通过递归的思想实现同样的过程。
